src/modules/Websocket.ts

```diff
@@ -0,0 +1,226 @@
+  private listeners: { [channel: string]: { [id: string]: (contents: any) => void } };
+
+  constructor(
+    timeout: number = DEFAULT_TIMEOUT_MS,
+    endpoint: string = DEFAULT_WS_ENDPOINT,
+    wsOrigin?: string,
+  ) {
+    this.wsOrigin = wsOrigin;
+    this.timeout = timeout;
+    this.endpoint = endpoint;
+  }
+
+  public async connect({
+    onClose = () => null,
+    onError = () => null,
+  }: {
+    onClose?: () => void,
+    onError?: (error: Error) => void,
+  } = {}): Promise<void> {
+    if (this.ws) {
+      throw new Error('Websocket already connected');
+    }
+
+    return this.reconnect({
+      onError,
+      onClose,
+    });
+  }
+
+  public async reconnect({
+    onClose = () => null,
+    onError = () => null,
+  }: {
+    onClose?: () => void,
+    onError?: (error: Error) => void,
+  } = {}): Promise<void> {
+    this.subscribedCallbacks = {};
+    this.listeners = {};
+    const options: any = {};
+    if (this.wsOrigin) {
+      options.origin = this.wsOrigin;
+    }
+
+    this.ws = new ws(
+      this.endpoint,
+      options,
+    );
+
+    this.ws.onClose = () => {
+      this.ws = null;
+      this.subscribedCallbacks = {};
+      this.listeners = {};
+      onClose();
+    };
+
+    this.ws.onMessage = (message: string) => {
+      let parsed: IncomingMessage;
+      try {
+        parsed = JSON.parse(message) as IncomingMessage;
+      } catch (error) {
+        onError(new Error(`Failed to parse websocket message: ${message}`));
+        return;
+      }
+
+      if (!Object.values(IncomingMessageType).includes(parsed.type)) {
+        onError(new Error(`Incomming message contained no type: ${message}`));
+        return;
+      }
+
+      if (parsed.type === IncomingMessageType.ERROR) {
+        onError(new Error(`Websocket threw error: ${(parsed as ErrorMessage).message}`));
+        return;
+      }
+
+      if (parsed.type === IncomingMessageType.SUBSCRIBED) {
+        const subscribedMessage = parsed as SubscribedMessage;
+
+        if (this.subscribedCallbacks[subscribedMessage.channel]) {
+          if (this.subscribedCallbacks[subscribedMessage.channel][subscribedMessage.id]) {
+            const callback = this.subscribedCallbacks
+              [subscribedMessage.channel][subscribedMessage.id];
+            delete this.subscribedCallbacks[subscribedMessage.channel][subscribedMessage.id];
+
+            callback(subscribedMessage.contents);
+          }
+        }
+
+        return;
+      }
+
+      if (parsed.type === IncomingMessageType.CHANNEL_DATA) {
+        const subscribedMessage = parsed as ChannelDataMessage;
+
+        if (this.listeners[subscribedMessage.channel]) {
+          if (this.listeners[subscribedMessage.channel][subscribedMessage.id]) {
+            const callback = this.listeners[subscribedMessage.channel][subscribedMessage.id];
+
+            callback(subscribedMessage.contents);
+          }
+        }
+
```

Comment:
nit: slightly less verbose is to say

```
const callback = this.listeners[subscribedMessage.channel][subscribedMessage.id];
if (callback) {
    callback(subscribeMessage.contents);
}
```

Or the equivalent logic for `this.subscribedCallbacks`